### PR TITLE
clear cached filemtime call, fix #12713

### LINKED ISSUE #12713: FileDependency is prone to fail in certain situations
### Minimal example:

config.php:

``` php
'components' => [
    'cache' => [
        'class' => 'yii\caching\FileCache',
    ],
    'volatileCache' => 'yii\caching\ArrayCache',
    ...
],
```

controller.php:

``` php
// insert appropriate namespace and use statements here
class TestController extends \yii\console\Controller
{
    private $_parsedContent;


    private function loadFromCache()
    {
        // Cache is volatile -> get('isModified') evaluates to false when first executed
        if (!Yii::$app->volatileCache->get('isModified')) {
            var_dump("Invalid Cache!");
            $this->_parsedContent = json_decode(Yii::getAlias('@app/someconfig.json'), true);
            Yii::$app->volatileCache->set('isModified', $value = true,
               $duration = 5, // invalidate after 5 seconds...
               new FileDependency(['fileName' => Yii::getAlias('@app/someconfig.json')])); // ...or when config file changes
        }

        return $this->_parsedContent;
    }

    public function actionIndex()
    {
        $this->loadFromCache();
        sleep(2);  // Modify the file during this sleep **outside** of php (in an editor)
        $this->loadFromCache(); // Expected output: Invalid Cache! (But thats not happening)
        sleep(4); //  Total waittime now 6 seconds -> Cache invalidates after 5
        $this->loadFromCache(); // Output: Invalid Cache! (But because of the regular timeout)
    }
}
```
### Problem Analysis

Interestingly enough the above example does work when using the `FileCache` component in favor of the `ArrayCache` component. Since I could not trace any direct call on `clearstatcache`, I assume that the stat cache is implicitly cleared by PHP after either of the calls to `file_put_contents`, `chmod` and `touch` inside `FileCache::setValue`.

Probably the easiest way to fix this bug is to explicitly call `clearstatcache` inside `FileDependency::generateDependencyData`, i.e. modify it as follows:

``` php
protected function generateDependencyData($cache)
{
    if ($this->fileName === null) {
        throw new InvalidConfigException('FileDependency::fileName must be set');
    }

    clearstatcache(false, Yii::getAlias($this->fileName));
    return @filemtime(Yii::getAlias($this->fileName));
}
```
### Additional info

Yii: 2.0.9
PHP: 7.0.8-0ubuntu0.16.04.2 (cli)

P.S. This issue may qualify for https://yourfirstpr.github.io/

### PR DESCRIPTION
**Pull request by SG5**
| Q | A |
| --- | --- |
| Is bugfix? | yes |
| New feature? | no |
| Breaks BC? | no |
| Tests pass? | yes/no |
| Fixed issues | #12713 |

### CODE REVIEW

**Comment by samdark** [comment]
Have you verified it helps?

**Comment by SG5** [comment]
yes, i've got 3 time `Invalid Cache!` message

**Comment by Coksnuss** [comment]
I can confirm that it works as well (as stated in my issue).
If the proposed fix is accepted, maybe the ticket can be moved to the 2.0.10 milestone?

File: framework/caching/FileDependency.php [hunk 1]
```diff
@@ -41,6 +41,7 @@ protected function generateDependencyData($cache)
             throw new InvalidConfigException('FileDependency::fileName must be set');
         }
 
+        clearstatcache(false, Yii::getAlias($this->fileName));
         return @filemtime(Yii::getAlias($this->fileName));
```
Comment (cebe) [inline framework/caching/FileDependency.php]: getAlias() should be called once and the result be stored in a variable.

**Review by cebe — CHANGES_REQUESTED** [review]

**Review by samdark — APPROVED** [review]

**Review by cebe — APPROVED** [review]

**Review by dynasource — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!

**Pull request merged**